23157 notes (#402)
bitcoin-core-review-club/website: _posts/2021-10-06-#23157.md

commit:
---

> We recently [looked at PR 22677](./22677), which cuts the circular dependency

Reviewer: I usually put the # sign in front of PR numbers but up to you
```suggestion
We recently [looked at PR #22677](./22677), which cuts the circular dependency
```

Author: Done!

between validation and txmempool. This week's PR was split off from that PR,
taking a slightly different approach to simplify the `CTxMemPool::check()`
function.

## Notes

- The `CTxMemPool` class is a data structure that stores the set of unconfirmed

>   transactions, along with various methods to update or query that set. As well
>   as the transactions themselves, the mempool stores metadata about the

Reviewer: ```suggestion
  transactions, along with various methods to update or query that set. In addition
  to the transactions themselves, the mempool stores metadata about the
```

  transactions (such as ancestor/descendant dependencies) that makes it highly
  optimized for various operations, such as constructing blocks.

- The mempool must only contain non-conflicting, valid transactions which

>   could theoretically be included in a future block. There are therefore

Reviewer: ```suggestion
  could theoretically be included in the next block. There are therefore
```

Reviewer: (stricter than future block. must be minable immediately)

Author: You're right!

  many [invariants](https://en.wikipedia.org/wiki/Class_invariant) such as:

  - a transaction may only appear in the mempool if all of its inputs are
    either in the UTXO set or are the outputs of other mempool transactions
  - no two transactions in the mempool may spend the same output

>   - the {mempool|ancestor} {fees|count|size} cached for the transaction must

Reviewer: I assume this is what it's supposed to say? Or self | descendant | ancestor?
```suggestion
  - the {descendant|ancestor} {fees|count|size} cached for the transaction must
```

Author: oops. Thank you!

    be the correct value

- There are several public methods that can alter the contents of the
  mempool. Those public methods can be called in many different
  circumstances, such as:

  - removing transactions when a block is connected

>   - expiring old transactions or removing transactions when limiting the

Reviewer: ```suggestion
  - expiring old transactions or evicting transactions by feerate when limiting the
```

Author: Done

  such as those fixed in [PR 2876](https://github.com/bitcoin/bitcoin/pull/2876) or
  [PR 5267](https://github.com/bitcoin/bitcoin/pull/5267).

- For that reason, [PR 2876](https://github.com/bitcoin/bitcoin/pull/2876)
  introduced a `CTxMemPool::check()` method, which asserts many of the
  mempool's invariants. Later, [PR
  5267](https://github.com/bitcoin/bitcoin/pull/5267) extended those checks.
  The checks are computationally expensive, so the `-checkmempool` command
  line option can control how frequently they're run.  By default, mempool
  checks are always disabled for mainnet, signet and testnet nodes, and always
  enabled for regtest nodes.

- This PR [significantly improves the performance of
  CTxMemPool::check()](https://github.com/bitcoin/bitcoin/pull/23157#issuecomment-932360429),
  and removes a dependency from that function to validation.h.

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?

2. Did you run the bench tests before and after the changes to
   `CTxMemPool::check()`? Were your results the same as reported in the PR?

> 3. Is it important for these checks to be fast? Why?

Reviewer: ```suggestion
3. Is it important for these checks to be fast? Why or why not?
```

4. What does the `UpdateCoins()` function do? What is it replaced with in
   this PR?

5. What was `waitingOnDependants` used for before this PR? Explain what
   [this while
loop](https://github.com/bitcoin/bitcoin/blob/446b70669/src/txmempool.cpp#L779-L790) was doing. Why can it now be removed?

> 6. How does the `GetSortedDepthAndScore()` function work? How does it ensure
>    that the vector that it returns doesn't contain any descendants before
>    their ancestors?

Reviewer: ```suggestion
6. How does the `GetSortedDepthAndScore()` function work? How does sorting
    by the number of ancestors ensure that ancestors precede descendants?
```

Author: I think this gives away too much of the answer. I've changed the wording of the second part of the question though: "Why does the returned vector not contain any descendants before their ancestors?"
